lib/sidekiq_long_running_job_logger.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+class SidekiqLongRunningJobLogger
+  attr_reader :thread
+
+  def initialize(stuck_sidekiq_job_minutes:)
+    @mutex = Mutex.new
+    @stop_requested = false
+    @stuck_sidekiq_job_minutes = stuck_sidekiq_job_minutes
+  end
+
+  def start
+    hostname = Discourse.os_hostname
+
+    @thread ||=
+      Thread.new do
+        loop do
+          break if self.stop_requested?
+
+          begin
+            current_long_running_jobs = Set.new
+
+            Sidekiq::Workers.new.each do |process_id, thread_id, work|
+              next unless process_id.start_with?(hostname)
+
+              if Time.at(work["run_at"]).to_i >= (Time.now - (60 * @stuck_sidekiq_job_minutes)).to_i
+                next
+              end
+
+              jid = work.dig("payload", "jid")
+              current_long_running_jobs << jid
+
+              next if @seen_long_running_jobs&.include?(jid)
+
+              if thread = Thread.list.find { |t| t["sidekiq_tid"] == thread_id }
+                Rails.logger.warn(<<~MSG)
+                Sidekiq job `#{work.dig("payload", "class")}` has been running for more than #{@stuck_sidekiq_job_minutes} minutes
+                #{thread.backtrace.join("\n")}
+                MSG
+              end
+            end
+
+            @seen_long_running_jobs = current_long_running_jobs
+
+            yield if block_given?
+          rescue => error
+            Discourse.warn_exception(
+              error,
+              message: "Unexpected error in SidekiqLongRunningJobChecker thread",
+            )
+          end
+
+          sleep(@stuck_sidekiq_job_minutes / 2)
```

Comment:
Why is this `@stuck_sidekiq_job_minutes / 2`? At first I found confusing the mix between the variable which represents minutes and the argument of `sleep` function which is in seconds. Then, I found confusing the division by 2.

But then, I realized that this has to be a low number to catch all long-running jobs, but not too long to give the jobs some time to advance.

So... with all this said, why is this not something constant like `sleep(5)`? Possible we could go even higher, maybe `sleep(30)` because we are only interested in jobs that have been stuck for at least a minute (assuming the unit of measure defined in the name of the setting).